base/reduce.jl

```diff
@@ -131,7 +131,13 @@ mapfoldr(f, op, v0, itr) = mapfoldr_impl(f, op, v0, itr, endof(itr))
 Like `mapfoldr(f, op, v0, itr)`, but using the first element of `itr` as `v0`. In general,
 this cannot be used with empty collections (see `reduce(op, itr)`).
 """
-mapfoldr(f, op, itr) = (i = endof(itr); mapfoldr_impl(f, op, f(itr[i]), itr, i-1))
+function mapfoldr(f, op, itr)
+    i = endof(itr)
```

Comment:
@pabloferz I considered that option over `isempty` but I found out that because of the following behavior I couldn't use it

```julia
julia> itr = Float64[]
0-element Array{Float64,1}

julia> i = endof(itr)
0

julia> done(itr, i)
false
```

It would only return `true` when `i == 1` for the empty array.